### ISSUE TITLE
Provide a custom hash() implementation on Tensor based on TensorImpl*

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#22884]
Currently, we use the default hash() implementation for Tensor, which is based on id() of the PyObject. However, it is possible for the id() of a tensor to change. Consider:

```
import torch
from torch.autograd import Function

class MyFunction(Function):
    @staticmethod
    def forward(ctx, x):
        y = x + 1
        ctx.save_for_backward(y)
        print(id(y), hash(y))
        return x

    @staticmethod
    def backward(ctx, grad_x):
        y = ctx.saved_tensors
        print(id(y), hash(y))
        return grad_x

MyFunction.apply(torch.zeros(1, requires_grad=True)).backward()
```

When run:

```
140588043707448 140588043707448
140588166083472 -6985488408845065285
```

What happened?  When we saved y to C++, we removed all Python references to the object. This caused the PyObject representing y to get deallocated. When we reallocate it later when it passes back into Python (via `saved_tensors`), it (obviously) gets a new address.

We can't "fix" id, but we can fix hash: just have it return a hash based on the address of the underlying TensorImpl, which is truly immutable.

cc @mrshenli @pietern

**Comment by colesbury** [comment]
How does that help you? If they don't have the same identity why make the hashes collide?

**Comment by ezyang** [comment]
@colesbury But they are the same tensor! (Sure, the PyObjects are different, but morally, they're still the same Tensor in the end.) Put differently, it is unexpected to the user that when they get back the saved variable, it's actually a different tensor.

**Comment by colesbury** [comment]
Again, what possible advantage can "fixing" the hash give? It can't be in a dictionary because that would have kept the original object alive.

**Comment by colesbury** [comment]
Also, iirc, it's not the same TensorImpl*: the whole saving for backwards unpacks the tensor. It's similar to how z.detach() (and z.data) won't give you the same object if you call it multiple times.

**Comment by colesbury** [comment]
https://github.com/pytorch/pytorch/blob/248336946e454ada1d00c51dc5c6073c504e4093/torch/csrc/autograd/python_function.cpp#L867-L874

**Comment by colesbury** [comment]
I think this is the key bit: SavedVariable::unpack always creates a new variable. It doesn't matter if you keep the original reference alive in Python.

https://github.com/pytorch/pytorch/blob/248336946e454ada1d00c51dc5c6073c504e4093/torch/csrc/autograd/saved_variable.cpp#L81-L86

**Comment by yf225** [comment]
Also we have to create a new Variable in SavedVariable::unpack, because the SavedVariable can be unpacked more than once:
```python
>>> a = torch.zeros(2, 2).fill_(3).requires_grad_()
>>> b = torch.zeros(2, 2).fill_(5).requires_grad_()
>>> c = a*b
# SavedVariable being saved:  3  3
#  3  3
# [ CPUFloatType{2,2} ]
# SavedVariable being saved:  5  5
#  5  5
# [ CPUFloatType{2,2} ]

>>> d = c.sum()
>>> d.backward(retain_graph=True)
# SavedVariable being unpacked:  3  3
#  3  3
# [ CPUFloatType{2,2} ]
# SavedVariable being unpacked:  5  5
#  5  5
# [ CPUFloatType{2,2} ]

>>> d.backward()
# SavedVariable being unpacked:  3  3
#  3  3
# [ CPUFloatType{2,2} ]
# SavedVariable being unpacked:  5  5
#  5  5
# [ CPUFloatType{2,2} ] 
```

**Comment by ezyang** [comment]
@colesbury You're right, I messed up my example. But the core problem still applies. Here is a better example:

```
import torch
import torch.utils.cpp_extension

cpp_source = r"""
#include <iostream>
#include <torch/extension.h>

torch::Tensor saved;
void put(torch::Tensor x) {
  saved = x;
}
torch::Tensor get() {
  return saved;
}
PYBIND11_MODULE(TORCH_EXTENSION_NAME, m) {
  m.def("put", &put, "put");
  m.def("get", &get, "get");
}
"""

module = torch.utils.cpp_extension.load_inline(
    name="foo",
    cpp_sources=[cpp_source],
)

x = torch.zeros(2)
print("x:", id(x))
module.put(x)
del x
y = torch.zeros(3)
print("y:", id(y))
print("x:", id(module.get()))
```

This prints:

```
x: 140365354752040
y: 140365354752040
x: 140365354775824
```

**Comment by ezyang** [comment]
> Again, what possible advantage can "fixing" the hash give? It can't be in a dictionary because that would have kept the original object alive.

That's fair enough (and I guess is a good reason to not fix this). In an ideal world you would have a weak reference to the key but we don't really have any way of doing that in the Python level.
